lib/libc/newlib/libc-hooks.c

```diff
@@ -270,3 +268,147 @@ __weak int *__errno(void)
+{
+	sys_mutex_unlock(&heap_mtx);
+}
+
+#else
+#include <sys/mempool.h>
+
+#define NEWLIB_LOCK_POOL_SIZE WB_UP(CONFIG_NEWLIB_LIBC_DYNAMIC_LOCK_MEM_SIZE)
+
+#define NEWLIB_LOCK_FRAG_SIZE WB_UP(sizeof(void *) + sizeof(struct k_mutex))
+
+K_APPMEM_PARTITION_DEFINE(z_malloc_lock_partition);
+#define NEWLIB_LOCK_SECTION K_APP_DMEM_SECTION(z_malloc_lock_partition)
+
+
+SYS_MEM_POOL_DEFINE(z_nl_lock_pool, NULL,
+		NEWLIB_LOCK_FRAG_SIZE, NEWLIB_LOCK_POOL_SIZE,
+		1, sizeof(void *), NEWLIB_LOCK_SECTION);
+
+static int newlib_lock_prepare(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	sys_mem_pool_init(&z_nl_lock_pool);
+
+	return 0;
+}
+
+SYS_INIT(newlib_lock_prepare, PRE_KERNEL_1,
+	CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_sinit_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_sfp_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_atexit_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_malloc_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_env_recursive_mutex);
+static LIBC_DATA SYS_SEM_DEFINE(nl_at_quick_exit_sem, 1, 1);
+static LIBC_DATA SYS_SEM_DEFINE(nl_tz_sem, 1, 1);
+static LIBC_DATA SYS_SEM_DEFINE(nl_dd_hash_sem, 1, 1);
+static LIBC_DATA SYS_SEM_DEFINE(nl_arc4random_sem, 1, 1);
+
+struct __lock {
+	void *pSemOrMtx;
+};
+
+struct __lock __lock___sinit_recursive_mutex = {&nl_sinit_recursive_mutex};
+struct __lock __lock___sfp_recursive_mutex = {&nl_sfp_recursive_mutex};
+struct __lock __lock___atexit_recursive_mutex = {&nl_atexit_recursive_mutex};
+struct __lock __lock___malloc_recursive_mutex = {&nl_malloc_recursive_mutex};
+struct __lock __lock___env_recursive_mutex = {&nl_env_recursive_mutex};
+
+struct __lock __lock___at_quick_exit_mutex = {&nl_at_quick_exit_sem};
+struct __lock __lock___tz_mutex = {&nl_tz_sem};
+struct __lock __lock___dd_hash_mutex = {&nl_dd_hash_sem};
+struct __lock __lock___arc4random_mutex = {&nl_arc4random_sem};
+
+
+void __retarget_lock_init(_LOCK_T *lock)
+{
+	lock = sys_mem_pool_alloc(&z_nl_lock_pool, NEWLIB_LOCK_FRAG_SIZE);
+
+	__ASSERT(lock, "failed to allocate memory for newlib lock");
+
+	(*lock)->pSemOrMtx = (void *) &((char *)lock)[sizeof(void *)];
+	sys_sem_init((struct sys_sem *) (*lock)->pSemOrMtx, 1, 1);
+}
+
+void __retarget_lock_close(_LOCK_T lock)
+{
+	sys_mem_pool_free(lock);
+}
+
+void __retarget_lock_acquire(_LOCK_T lock)
+{
+	sys_sem_take((struct sys_sem *) lock->pSemOrMtx, K_FOREVER);
+}
+
+int __retarget_lock_try_acquire(_LOCK_T lock)
+{
+	if (sys_sem_take((struct sys_sem *) lock->pSemOrMtx,
+		K_NO_WAIT) == 0) {
+		return 1;
+	}
+	return 0;
+}
+
+void __retarget_lock_release(_LOCK_T lock)
+{
+	sys_sem_give((struct sys_sem *) lock->pSemOrMtx);
+}
+
+
+void __retarget_lock_init_recursive(_LOCK_T *lock)
+{
+	lock = sys_mem_pool_alloc(&z_nl_lock_pool, NEWLIB_LOCK_FRAG_SIZE);
+
+	__ASSERT(lock, "failed to allocate memory for newlib lock");
+
+	(*lock)->pSemOrMtx = (void *) &((char *)lock)[sizeof(void *)];
```

Comment:
I'm not understanding this function.  You're taking a lock pointer as an argument (where is that defined?), but then throwing that value away and replacing it with a new heap block (which may be null, and is unchecked), then dereferencing whatever pointer happened to be stored in that uninitialized heap block to store a pointer into the same block?

I'm guessing that what you really want to be doing is allocating a block containing just the sem/mutex union and assigning that through the opaque pointer you're being passed?

Why is there a header containing pSemOrMtx if the pointer always points to the byte after its own address?  Why not just cast the struct address in the first place?

Reply:
> You're taking a lock pointer as an argument (where is that defined?), but then throwing that value away and replacing it with a new heap block

Yeah this is confusing me too, some more detail on the intention here would be helpful, maybe leave a comment if this is truly correct (although right now it looks like the allocated lock simply leaks)

Reply:
Fixed the possible null pointer dereference

> I'm guessing that what you really want to be doing is allocating a block containing just the sem/mutex union and assigning that through the opaque pointer you're being passed?

> Why is there a header containing pSemOrMtx if the pointer always points to the byte after its own address? Why not just cast the struct address in the first place?

Your right, i've changed that.